### PR TITLE
add doc test in yaml

### PR DESCRIPTION
**Pull request by hantek**
A fix for #3103

### CODE REVIEW

**Comment by nouiz** [comment]
You need to rebase this PR. Travis won't run on it as there is conflict.

**Comment by hantek** [comment]
A fie for #3103

**Comment by hantek** [comment]
The travis said "permission denied." Can I use chmod or something like that in a travis yaml file?

The command "if [ ${DOC} = "1" ]; then doc/scripts/docgen.py --nopdf; else theano-nose -v $PART; fi" exited with 126.

0.00s$ if [ ${DOC} = "1" ]; then doc/scripts/docgen.py --test; else theano-nose -v $PART; fi

/home/travis/build.sh: line 45: doc/scripts/docgen.py: Permission denied

**Comment by nouiz** [comment]
python doc/scripts/docgen.py

I'm not sure you can use chmod and it will be easier this way.

**Comment by hantek** [comment]
I need to install some dependencies for the docgen. Some of them requires a sudo permission. How do I get that in a travis machine?

**Comment by nouiz** [comment]
What do you need to install? the machine allow to install via apt-get, but
we use conda. Check the travis config file for the line that install via
conda.

**Comment by hantek** [comment]
It is epydoc. I checked their website and they seem not having a apt-get
installation option. So I am trying to compile from the source code and
install.

On Friday, November 13, 2015, Frédéric Bastien <notifications@github.com>
wrote:

> What do you need to install? the machine allow to install via apt-get, but
> we use conda. Check the travis config file for the line that install via
> conda.
>
> On Fri, Nov 13, 2015 at 4:33 PM, Zhouhan LIN <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> wrote:
>
> > I need to install some dependencies for the docgen. Some of them requires
> > a sudo permission. How do I get that in a travis machine?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/Theano/Theano/pull/3631#issuecomment-156562061>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3631#issuecomment-156562654>.
>


-- 
LIN, Zhouhan
Department of Computer Science and Operational Research
University of Montreal

Phone: +01-514-553-6037
Email: lin.zhouhan@gmail.com
Homepage:http://zhouhanlin.webs.com/

**Comment by hantek** [comment]
The link to their site is http://epydoc.sourceforge.net/manual-install.html

On Friday, November 13, 2015, Zhouhan LIN <lin.zhouhan@gmail.com> wrote:

> It is epydoc. I checked their website and they seem not having a apt-get
> installation option. So I am trying to compile from the source code and
> install.
>
> On Friday, November 13, 2015, Frédéric Bastien <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> What do you need to install? the machine allow to install via apt-get, but
>> we use conda. Check the travis config file for the line that install via
>> conda.
>>
>> On Fri, Nov 13, 2015 at 4:33 PM, Zhouhan LIN <notifications@github.com>
>> wrote:
>>
>> > I need to install some dependencies for the docgen. Some of them
>> requires
>> > a sudo permission. How do I get that in a travis machine?
>> >
>> > —
>> > Reply to this email directly or view it on GitHub
>> > <https://github.com/Theano/Theano/pull/3631#issuecomment-156562061>.
>> >
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/Theano/Theano/pull/3631#issuecomment-156562654>.
>>
>
>
> --
> LIN, Zhouhan
> Department of Computer Science and Operational Research
> University of Montreal
>
> Phone: +01-514-553-6037
> Email: lin.zhouhan@gmail.com
> <javascript:_e(%7B%7D,'cvml','lin.zhouhan@gmail.com');>
> Homepage:http://zhouhanlin.webs.com/
>
>

-- 
LIN, Zhouhan
Department of Computer Science and Operational Research
University of Montreal

Phone: +01-514-553-6037
Email: lin.zhouhan@gmail.com
Homepage:http://zhouhanlin.webs.com/

**Comment by abergeron** [comment]
I would rather remove support for epydoc in docgen.py.

**Comment by nouiz** [comment]
I also think we don't use it anymore. Can you just delete all what is
related to epydoc? The import and the related code.

**Comment by hantek** [comment]
I added this line in the "install" key, but it seems not working... Travis still reports Sphinx not found. But isn't it the way to install Sphinx?

  - pip install -U Sphinx --no-deps

**Comment by hantek** [comment]
Travis is passing.

Looking at the Travis log, there are still "failed tests." (like attached) Should we fix that? And if those "fails" are considered to be "fail," how to set Travis yaml file not to ignore them?

Document: library/gof/graph
---------------------------
**********************************************************************
File "library/gof/graph.txt", line 18, in default
Failed example:
    z.eval({x : 16.3, y : 12.1})
Expected:
    array(28.4)
Got:
    array(28.399999999999999)
**********************************************************************

File: .travis.yml [hunk 1]
```diff
@@ -56,7 +61,9 @@ script:
   # Move to the path of the installed version
   - cd $(python -c 'import theano; import os; print(os.path.split(theano.__file__)[0])')
   - echo "$PART"
-  - theano-nose -v $PART
+  - cd -; cd Theano
+  - if [[ $DOC == "1" ]]; then python doc/scripts/docgen.py --nopdf; else theano-nose -v $PART; fi
+  - if [[ $DOC == "1" ]]; then python doc/scripts/docgen.py --test; else theano-nose -v $PART; fi
```
Comment (nouiz) [inline .travis.yml]: this trigger 2 time Theano tests when not in DOC mode. Just remove the else part on this line.

File: doc/scripts/docgen.py [hunk 2]
```diff
@@ -116,7 +64,7 @@ def mkdir(path):
         #               '-o', 'api']
         #cli()
         ## So we use this instead
-        os.system("epydoc --config %s/doc/api/epydoc.conf -o api" % throot)
+        #os.system("epydoc --config %s/doc/api/epydoc.conf -o api" % throot)
```
Comment (nouiz) [inline doc/scripts/docgen.py]: Delete this section.

**Comment by nouiz** [comment]
I don't know why, but travis didn't ran on the last commit. I'll open and close this PR totry to trigger it.

Before fixing the tests, did you rebase as we discussed? I already fixed some of them in another PR that was done after you started to work on this.

Also, we should make travis fail when one or more tests fail. Work on this before fixing the tests.

File: doc/scripts/docgen.py (hunk 2 above)
Comment (nouiz) [inline doc/scripts/docgen.py]: I mean, delete the commented if and code commented code inside the if.

**Comment by hantek** [comment]
Yes I have rebased this PR on the last commit at Theano/master 10 days ago. Will work on those failing tests then.

**Comment by nouiz** [comment]
Before fixing the tests, make sure that travis fail when some tests fail. We want travis to fail as soon as 1 doctest fail.

**Comment by hantek** [comment]
Travis is passing. I think all issues have been solved now.

**Comment by nouiz** [comment]
If some doctest fail, this don't make travis fail. We need that. See:

https://travis-ci.org/Theano/Theano/jobs/93390169

it have failing doc test, but have passed.

**Comment by hantek** [comment]
Hi Fred,

I am getting this error by sphinx, which I don't know how to solve:

        Warning, treated as error:
        /u/linzhou/Theano/doc/library/sandbox/cuda/op.txt:22: WARNING: error while formatting arguments for theano.sandbox.cuda.blas.batched_dot: 'BatchedDotOp' object has no attribute '__bases__'

It is using "autofunction" in the file doc/library/sandbox/cuda/op.txt, I googled a bit and I find an issue in sphinx repo suggesting not to care about this kind of warning. Do you have any suggestions?

Zhouhan

**Comment by hantek** [comment]
Also, in the file Theano/doc/tutorial/profiling.txt, there are some test codes like:

         39           profile = theano.compile.ProfileStats()
         40           f = theano.function(..., profile=profile)
         41           g = theano.function(..., profile=profile)
         42           ...
         43           profile.print_summary()

I suppose this code is not intended for executing, but the doctest is executing them anyway. So how to exclude them from those tests  being actually executed? (But still be presented in the form of code in the generated documentation.)

**Comment by hantek** [comment]
I suspect the "theano.sandbox.cuda.blas.batched_dot" there should be a class name, not a class object. Thus it should be BatchedDotOp instead. In that way the former "has no attribute" error can be eliminated.

**Comment by nouiz** [comment]
I think it is better to have batched_dot being documented. Otherwise, it tell the user to instantiate that op before using it. We could rename the current batched_dot to batched_dot_ and create a function named batched_dot that call it.
```
batched_dot_ = BatchedDotOp()
def batched_dot(...):
 """
 """
 return batched_dot_(...)
```

To find how to skip some tests, see: http://stackoverflow.com/questions/1809037/python-doctest-skip-entire-block

For the profiling file, this file got updated. So rebase and this will probably get fixed.

**Comment by nouiz** [comment]
There is this pep8 error:

/home/travis/build/Theano/Theano/theano/compile/function.py:53:1: W293 blank line contains whitespace

There is still 4 failures and this didn't make travis crash.

**Comment by hantek** [comment]
Hi Fred,

Now every warinings in documentations are solved. And all the errors in doctest are solved, too. 

Besides that, I find that changing the sphinx flag to "-M" doesn't solve the doctest failure problem: If there were some failures in doctest, it doesn't cause test test to fail. (at least on my local machine, but I guess it should be the same on travis machines.) 

A better way to make doctest alert when comign across doctest failures?

Zhouhan

**Comment by nouiz** [comment]
I'll merge now even if one part is missing, making travis fail when there is a documentation problem. I'll let this to another PR.

**Pull request merged**